pkg/admission/bootstrap.go

```diff
@@ -0,0 +1,256 @@
+			return
+		}
+		s.Client, err = client.New(nil, client.Options{})
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	awcs, err := s.whConfigs()
+	if err != nil {
+		return nil, err
+	}
+
+	// if dryrun, return the AdmissionWebhookConfiguration in yaml format.
+	if dryrun {
+		return s.genYamlConfig(awcs)
+	}
+
+	for i := range awcs {
+		err = s.CertProvisioner.Sync(awcs[i])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	svc := s.service()
+	objects := awcs
+	if svc != nil {
+		objects = append(awcs, svc)
+	}
+	for _, obj := range objects {
+		err = createOrReplace(s.Client, obj)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return nil, nil
+}
+
+// createOrReplace creates the object if it doesn't exist;
+// otherwise, it will replace it.
+func createOrReplace(client client.Client, obj runtime.Object) error {
+	err := client.Create(context.Background(), obj)
+	if apierrors.IsAlreadyExists(err) {
+		return client.Update(context.Background(), obj)
+	}
+	return err
+}
+
+// genYamlConfig generates yaml config for admissionWebhookConfiguration
+func (s *Server) genYamlConfig(whConfigs []runtime.Object) ([]byte, error) {
+	var buf bytes.Buffer
+	firstObject := true
+	for _, awc := range whConfigs {
+		if !firstObject {
+			_, err := buf.WriteString("---")
+			if err != nil {
+				return nil, err
+			}
+		}
+		firstObject = false
+		b, err := yaml.Marshal(awc)
+		if err != nil {
+			return nil, err
+		}
+		_, err = buf.Write(b)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return buf.Bytes(), nil
+}
+
+func (s *Server) annotations(t WebhookType) map[string]string {
+	anno := map[string]string{}
+	for path, handler := range s.registry {
+		if handler.t != t {
+			continue
+		}
+		if s.Secret != nil {
+			key := writer.SecretCertProvisionAnnotationKeyPrefix + handler.name
+			anno[key] = s.Secret.String()
+		} else {
+			key := writer.FSCertProvisionAnnotationKeyPrefix + handler.name
+			anno[key] = path
+		}
+	}
+	return anno
+}
+
+// whConfigs creates a mutatingWebhookConfiguration and(or) a validatingWebhookConfiguration based on registry.
+// For the same type of webhook configuration, it generates a webhook entry per endpoint.
+func (s *Server) whConfigs() ([]runtime.Object, error) {
+	if s.Host != nil && s.Service != nil {
+		return nil, errors.New("URL and Service can't be set at the same time")
+	}
+
+	mutatingWebhooks, validatingWebhooks := []v1beta1.Webhook{}, []v1beta1.Webhook{}
+	var mutatingWebhookConfig *admissionregistration.MutatingWebhookConfiguration
+	var validatingWebhookConfig *admissionregistration.ValidatingWebhookConfiguration
+	for path, handler := range s.registry {
+		wh := admissionregistration.Webhook{
+			Name:              handler.name,
+			Rules:             handler.rules,
+			FailurePolicy:     handler.failurePolicy,
+			NamespaceSelector: handler.namespaceSelector,
+			ClientConfig: admissionregistration.WebhookClientConfig{
+				// assigning an empty CABundle, since this will be updated.
```

Comment:
maybe add `TODO(xxx)` to highlight in editor.

Reply:
Here doesn't meant to be a TODO. I have rephrased it a bit.